Save, delete and update tags locally and remotely

ctarda opened the pull request:
Implements part of #8232 

Another early PR to check the general direction I've taken. 😊 

This one contains updates to the `PostTagService` and the Rest and XMLRPC taxonomies.

Saving and deleting seem to work fine, updating tags is a bit broken though. I am still looking into it, but in the meantime, more work for you, @aerych ! 😛

aerych commented:
> in the meantime, more work for you, @aerych ! 😛

Bring it! :D

ctarda commented:
Hi again! 👋  Whenever you look into this, could you please pay some special attention to [this commit](https://github.com/wordpress-mobile/WordPress-iOS/pull/8305/commits/3cfd7134a42fbb01554dd4426b6a0c8f63f8963c)?

It seems like when I update a tag, the RemoteTag I get back from the taxonomy service has an identifier that does not match the identifier of the RemoteTag I send. So what I do in that specific commit is update the core data managed object with the said remote tag. 

Is that what I am supposed to do? 🤔

ctarda commented:
One more thing: I am trying to test the XMLRPC implementation of the `TaxonomyServiceRemote`, but I can't find to do it. I have been playing with the configuration of a couple of my self-hosted blogs, but I don't hit that class ☺️

aerych commented on WordPress/Classes/ViewRelated/Blog/SiteManagement/SiteTagsViewController.swift:
```diff
@@ -231,12 +231,20 @@ extension SiteTagsViewController {
+        tagsService.delete(tag, for: blog, success: {
+            self.refreshControl?.endRefreshing()
+            self.tableView.reloadData()
+        }, failure: nil)
```
If the call fails would we still want to call `self.refreshControl?.endRefreshing()`?

aerych commented on WordPress/Classes/ViewRelated/Blog/SiteManagement/SiteTagsViewController.swift:
```diff
@@ -231,12 +231,20 @@ extension SiteTagsViewController {
+        tagsService.save(tag, for: blog, success: { [weak self] tag in
+            self?.refreshControl?.endRefreshing()
+            self?.tableView.reloadData()
+        }, failure: nil)
```
Here too

aerych commented on WordPress/Classes/Services/PostTagService.m:
```diff
@@ -118,16 +120,38 @@ - (void)searchTagsWithName:(NSString *)nameQuery
+        if (success) {
+            success();
+        }
+    } failure:failure];
```
It would be nice if we could call the `handleError:forBlog:withFailure` to log the error description.

aerych commented:
Oh two more things I forgot to mention (sorry!).
- I'd noticed when creating a new tag the description did not appear if I viewed the tag details after saving.  My steps were:
Tap + button > Enter tag name and description > Tap Back button.  Wait a moment.  Tap the new tag to view.  The name was present but the description was blank.  I could still edit the tag to add a description successfully.  
- When deleting a tag, the tag remains visible in the list until the delete operation completes.  This could be a little confusing, especially on a slow connection.  It might be nice to optimistically delete the tag rather than wait.

ctarda commented on WordPressKit/WordPressKit/TaxonomyServiceRemote.h:
```diff
@@ -46,6 +46,20 @@ NS_ASSUME_NONNULL_BEGIN
+ Delete a tag with the site.
+ */
+- (void)deleteTag:(RemotePostTag *)tag
+		  success:(nullable void (^)(RemotePostTag *tag))success
```
We don't need it, I kind of left it there because it made the API look more symmetric, which obviously, is not a reason good enough. I'll remove it.

ctarda commented on WordPressKit/WordPressKit/TaxonomyServiceRemoteREST.m:
```diff
@@ -107,6 +107,42 @@ - (void)createTag:(RemotePostTag *)tag
+	[self updateTaxonomyWithType:TaxonomyRESTTagIdentifier
+					  parameters:parameters success:^(NSDictionary * _Nonnull responseObject) {
+						  RemotePostTag *receivedTag = [self remoteTagWithJSONDictionary:responseObject];
+						  if (success) {
```
Ouch, that one hurt! 😝

ctarda commented:
Howdy, @aerych ! 👋 

I think this is ready for another round of review. I believe I addressed all your comments, with one "but" 😛 

With regards to the tabs and spaces, what I did was change Xcode's editor configuration, and then command + A and control + I in the offending files, but to be honest, I don't know if that's enough, and I also assume it will make the diff slightly larger. Anyway, I isolated that change in the very last commit, so we can easily revert and pretend I didn't break anything if necessary 😊

aerych commented on WordPressKit/WordPressKit/TaxonomyServiceRemoteXMLRPC.m:
```diff
@@ -102,6 +104,39 @@ - (void)createTag:(RemotePostTag *)tag
+    NSMutableDictionary *extraParameters = [NSMutableDictionary dictionary];
+    [extraParameters setObject:tag.tagID ?: [NSNull null] forKey:TaxonomyXMLRPCIDParameter];
+    [extraParameters setObject:tag.name ?: [NSNull null] forKey:TaxonomyXMLRPCNameParameter];
+    [extraParameters setObject:tag.description ?: [NSNull null] forKey:TaxonomyXMLRPCDescriptionParameter];
```
Should `tag.description` be `tag.tagDescription`?   When I inspect the parameters being passed I see something like `"<RemotePostTag: 0x60400065b750> Tag6[14]";` for its value.

aerych commented on WordPressKit/WordPressKit/TaxonomyServiceRemoteXMLRPC.m:
```diff
@@ -102,6 +104,39 @@ - (void)createTag:(RemotePostTag *)tag
+          failure:(nullable void (^)(NSError *error))failure
+{
+    NSMutableDictionary *extraParameters = [NSMutableDictionary dictionary];
+    [extraParameters setObject:tag.tagID ?: [NSNull null] forKey:TaxonomyXMLRPCIDParameter];
```
Sort of.  This does add the `term_id` but where it appears matters.  With this change, if you were to inspect `xmlrpcParameters` in `editTaxonomyWithType` you'd see something like:
```
(lldb) po xmlrpcParameters
<__NSArrayM 0x60400065bb70>(
0,
username,
password,
{
    term_id = 14;
    description = "<RemotePostTag: 0x60400065b750> Tag6[14]";
    name = Tag6;
    taxonomy = "post_tag";
}
)
```
However, the docs for the XML-RPC method `wp.editTerm` specify that the term ID should be passed after the password.  The params should look something like this:
```
(lldb) po xmlrpcParameters
<__NSArrayM 0x60400065bb70>(
0,
username,
password,
14,
{
    description = "<RemotePostTag: 0x60400065b750> Tag6[14]";
    name = Tag6;
    taxonomy = "post_tag";
}
)
```

aerych commented on WordPressKit/WordPressKit/TaxonomyServiceRemoteXMLRPC.m:
```diff
@@ -102,6 +104,39 @@ - (void)createTag:(RemotePostTag *)tag
+          failure:(nullable void (^)(NSError *error))failure
+{
+    NSMutableDictionary *extraParameters = [NSMutableDictionary dictionary];
+    [extraParameters setObject:tag.tagID ?: [NSNull null] forKey:TaxonomyXMLRPCIDParameter];
```
Same issue here as in `editTag:success:failure:` with where the term_id is passed.

aerych commented on WordPressKit/WordPressKit/TaxonomyServiceRemoteXMLRPC.m:
```diff
@@ -89,6 +90,7 @@ - (void)createTag:(RemotePostTag *)tag
 {
     NSMutableDictionary *extraParameters = [NSMutableDictionary dictionary];
     [extraParameters setObject:tag.name ?: [NSNull null] forKey:TaxonomyXMLRPCNameParameter];
+    [extraParameters setObject:tag.description ?: [NSNull null] forKey:TaxonomyXMLRPCDescriptionParameter];
```
Should tag.description be tag.tagDescription here as well?

ctarda commented:
Hiya, @aerych ! I just pushed another update!

ctarda commented:
Hi @aerych ! I think the XML-RPC side of things is finally working as expected! I think this is ready for another round of review 😄
